fv3gfs/util/halo_updater.py

```diff
@@ -0,0 +1,263 @@
+            comm: communicator to post network messages
+            numpy_like_module: module implementing numpy API
+            quantities_x: quantities to exchange along the x axis.
+                          Length must match y quantities.
+            quantities_y: quantities to exchange along the y axis.
+                          Length must match x quantities.
+            boundaries: informations on the exchange boundaries.
+            tag: network tag (to differentiate messaging) for this node.
+            n_halo_points: size of the halo to exchange.
+            optional_timer: timing of operations.
+        
+        Returns:
+            HaloUpdater ready to exchange data.
+        """
+        timer = optional_timer if optional_timer is not None else NullTimer()
+
+        packed_buffers: List[Tuple[int, HaloDataTransformer]] = []
+        for boundary in boundaries:
+            for specification_x, specification_y in zip(
+                specifications_x, specifications_y
+            ):
+                buffer = cls._get_packed_buffer(
+                    packed_buffers, numpy_like_module, boundary
+                )
+                buffer.add_vector_specification(
+                    specification_x,
+                    specification_y,
+                    boundary.send_slice(specification_x, n_halo_points),
+                    boundary.send_slice(specification_y, n_halo_points),
+                    boundary.n_clockwise_rotations,
+                    boundary.recv_slice(specification_x, n_halo_points),
+                    boundary.recv_slice(specification_y, n_halo_points),
+                )
+
+        for _to_rank, buffer in packed_buffers:
+            buffer.compile()
+
+        return cls(comm, tag, packed_buffers, timer)
+
+    @classmethod
+    def _get_packed_buffer(
+        cls,
+        packed_buffer: List[Tuple[int, HaloDataTransformer]],
+        numpy_like_module: NumpyModule,
+        boundary: Boundary,
+    ) -> HaloDataTransformer:
+        """Returns the correct packed_buffer from the list create it first if need be.
+
+        Args:
+            packed_buffer: list of Tuple [target_rank_to_send_data, HaloDataTransformer]
+            numpy_like_module: module implementing numpy API
+            boundary: information on the exchange boundary
+
+        Returns:
+            Correct HaloDataTransformer for target rank
+        """
+        to_rank_packed_buffer = [x for x in packed_buffer if x[0] == boundary.to_rank]
+        assert len(to_rank_packed_buffer) <= 1
+        if len(to_rank_packed_buffer) == 0:
+            packed_buffer.append(
+                (
+                    boundary.to_rank,
+                    HaloDataTransformer.get_from_numpy_module(numpy_like_module),
+                )
+            )
+            to_rank_packed_buffer = [
+                x for x in packed_buffer if x[0] == boundary.to_rank
+            ]
+        return to_rank_packed_buffer[0][1]
+
+    def do(
```

Comment:
nit: would suggest naming this `update` so even if the instance name is not clear it's obvious that this updates its arguments in place. It's also a bit of a keyword for "halo update".

Reply:
Done